lib/seek/fair_data_station/writer.rb

```diff
@@ -1,24 +1,44 @@
 module Seek
   module FairDataStation
+    class ExternalIdMismatchException < RuntimeError; end;
     class Writer
       def construct_isa(datastation_inv, contributor, projects, policy)
         reset_data_file_cache
-        investigation = build_investigation(datastation_inv, contributor, policy, projects)
+        investigation = build_investigation(datastation_inv, contributor, projects, policy)
 
         datastation_inv.studies.each do |datastation_study|
           study = build_study(datastation_study, contributor, policy, investigation)
           datastation_study.observation_units.each do |datastation_observation_unit|
-            observation_unit = build_observation_unit(datastation_observation_unit, contributor, policy, projects,
+            observation_unit = build_observation_unit(datastation_observation_unit, contributor, projects, policy,
                                                       study)
             datastation_observation_unit.samples.each do |datastation_sample|
-              sample = build_sample(datastation_sample, contributor, policy, projects)
-              if sample.valid?
-                observation_unit.samples << sample
-              else
-                Rails.logger.error("Invalid sample during fair data station import #{sample.errors.full_messages.inspect}")
+              sample = build_sample(datastation_sample, contributor, projects, policy, observation_unit)
+              datastation_sample.assays.each do |datastation_assay|
+                build_assay(datastation_assay, contributor, projects, policy, sample, study)
               end
+            end
+          end
+        end
+
+        investigation
+      end
+
+      def update_isa(investigation, datastation_inv, contributor, projects, policy)
```

Comment:
I understand that the current `update_isa` method only supports fair_data_station resource types derived from SEEK resources. In the future, could this method be adapted for use with SEEK ISA elements as well?  A good use case could be creating and updating a whole ISA through the API.

Reply:
yes, although thinking more in terms of supporting RO-Crates. The FAIR data station TTL is very similar to what an RO-Crate may look like.
For SEEK to SEEK migration the UUID's may be safer to use than the external identifiers, it's what they were originally added for.